remote_execution/oss/re_grpc/src/client.rs

```diff
@@ -466,77 +577,240 @@ impl REClient {
         Ok(stream.boxed())
     }
 
+    pub async fn check_needed_blobs(
+        &self,
+        request: &UploadRequest,
+    ) -> anyhow::Result<HashSet<TDigest>> {
+        let mut cas_client = self.grpc_clients.cas_client.clone();
+        let mut needed_blobs = HashSet::new();
+        if let Some(blobs_with_digest) = &request.inlined_blobs_with_digest {
+            for blobs in blobs_with_digest.chunks(100) {
+                let missing_blobs = cas_client
+                    .find_missing_blobs(FindMissingBlobsRequest {
+                        instance_name: INSTANCE_NAME.to_string(),
+                        blob_digests: blobs.map(|b| tdigest_to(b.digest.clone())),
+                    })
+                    .await?;
+                let resp: FindMissingBlobsResponse = missing_blobs.into_inner();
+                for blob in resp.missing_blob_digests {
+                    needed_blobs.insert(tdigest_from(blob));
+                }
+            }
+        }
+
+        if let Some(files_with_digest) = &request.files_with_digest {
+            for files in files_with_digest.chunks(100) {
+                let missing_blobs = cas_client
+                    .find_missing_blobs(FindMissingBlobsRequest {
+                        instance_name: INSTANCE_NAME.to_string(),
+                        blob_digests: files.map(|b| tdigest_to(b.digest.clone())),
+                    })
+                    .await?;
+                let resp: FindMissingBlobsResponse = missing_blobs.into_inner();
+                for blob in resp.missing_blob_digests {
+                    needed_blobs.insert(tdigest_from(blob));
+                }
+            }
+        }
+
+        anyhow::Ok(needed_blobs)
+    }
```

Comment:
How would `Uploader` have context on what other clients have uploaded already? Additionally, I'll note that without this snippet, remote execution is currently unusable for my workload.

Reply:
> How would Uploader have context on what other clients have uploaded already? Additionally, I'll note that without this snippet, remote execution is currently unusable for my workload.

It calls `get_digests_ttl`, here:

https://github.com/facebook/buck2/blob/f3b83eeba0f395359456c250bfddc951757ddec4/app/buck2_execute/src/re/uploader.rs#L102-L107

It then only uploads things that are about to expire.

However, the problem is that in this client, this is implemented as ... a stub:

https://github.com/facebook/buck2/blob/f3b83eeba0f395359456c250bfddc951757ddec4/remote_execution/oss/re_grpc/src/client.rs#L566-L578

(I didn't realize, my bad, but _that_ is the bit that needs changing)

Specifically, what would need to be changed here is getting this to make a call to find_missing, then return a more realistic TTL for blobs that are found to not be missing. At the very least, that TTL would have to be one that Uploader would be happy with (right now, it expects a TTL of 600 seconds). Then, the upload would not attempt to re-upload things that currently exist.

The RE API doesn't actually provide a way to discover the TTL of blobs (the `get_digests_ttl` method we use internally does, but of course that's not something you can use here). Production RE backends should normally extend the TTL of blobs when they are queried (otherwise you've got a race window no matter what when you call find_missing), but it's also not possible to discover this.

The current design does force you to make a decision as to what you think is the TTL when your backend says the blob isn't missing. I think a reasonable way to solve this would be make both the ttl_wanted in Uploader and the ttl returned by `get_digests_ttl` configurable, at which point we can probably set them to rather low defaults for OSS builds.

Reply:
That makes sense, and explains my confusion! I'll add some logic there.

Reply:
I moved the logic, seems to work :+1: